packages/data/src/hypoCompress.ts

```diff
@@ -0,0 +1,406 @@
+import { isEqual } from "lodash";
+import { HYPO_PLAYER_NAMES, MAX_PLAYERS, MIN_PLAYERS } from "./constants";
+import { getVariant, getVariantByID } from "./gameData";
+import { parseIntSafe } from "./utils";
+
+interface DeckCard {
+  suitIndex: number;
+  rank: number;
+}
+
+interface Action {
+  type: number;
+  target: number;
+  value?: number;
+}
+
+interface Options {
+  variant: string | undefined;
+}
+
+interface MinMax {
+  min: number;
+  max: number;
+}
+
+export interface GameJSON {
+  players: string[];
+  deck: DeckCard[];
+  actions: Action[];
+  options: Options;
+  characters: string[];
+  id: number;
+  notes: [];
+  seed: string;
+}
+
+const BASE62 = "abcdefghijklmnopqrstuvwxyz0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+
+/**
+ * Compresses a string representing a GameJSON object.
+ * Returns null if the compression fails.
+ *
+ * The resulting string is composed of three substrings separated by comma.
+ * The first substring represents the number of players and the deck.
+ * The second substring represents the actions.
+ * The third substring is the ID of the variant.
+ *
+ * Finally hyphens are added to the string to make URL text wrap when posting.
+ *
+ * @param JSONString A string representing a hypo JSON
+ */
+export function shrink(JSONString: string): string | null {
+  let gameDataJSON: GameJSON;
+  try {
+    gameDataJSON = JSON.parse(JSONString) as GameJSON;
+  } catch {
+    return null;
+  }
+
+  const compressed = gameJSONCompress(gameDataJSON);
+  if (compressed === null) {
+    return null;
+  }
+
+  // Check if compression succeeded
+  const decompressed = expand(compressed);
+  if (decompressed === null) {
+    return null;
+  }
+  try {
+    const decompressedJSON = JSON.parse(decompressed) as GameJSON;
+    if (!isEqual(gameDataJSON, decompressedJSON)) {
+      return null;
+    }
+  } catch {
+    return null;
+  }
+
+  return compressed;
+}
+
+/**
+ * Decompresses a string into a GameJSON object.
+ * Returns null if decompression fails.
+ *
+ * @param data The compressed string
+ */
+export function expand(data: string): string | null {
+  // Remove all hyphens from URL
+  const normal = data.replace(/-/g, "");
+
+  // The compressed string is composed of 3 substrings, separated by comma
```

Comment:
should be "commas" instead of "comma"

Reply:
Actually it should be commata (greek word, plural) but I'll settle for commas.

Reply:
=p